File: C/stack.c
#include <stdio.h>
#include <stdlib.h>

int top = -1; //Since the stack is not yet created, initialized the top of the stack as -1 since it doesnt exist yet.

void push(int la[], int x) //Function to add elements on top of the stack
{

    if (top >= x - 1)//If the stack is full ie the top value is equal to the specified no. of elements - 1, the no elements can be added.
    {
        printf("Stack is overflow\n\n");//Condition is called overflow.
    }
    else//If stack is not full
    {
        int item;
        printf("Enter a value to be pushed : ");
        scanf("%d", &item);
        top++;          //Before adding the item first increment top variable so element is added at the next position.
        la[top] = item; //Added the item in the respective place.
    }
    printf("\n");
}

void pop(int la[]) //Function to remove elements from top of the stack.
{

    if (top <= -1)//If the top value is -1, this means that stack is empty and there is nothing to delete. This is called underflow condtion.
    {
        printf("Stack is underflow\n\n");
    }
    else //If the stack is not empty.
    {
        printf("The popped element is " "%d" "\n\n", la[top]);//Print the element which is deleted.
        top--;//Decrement the top value since the previous top item is removed
    }
}

int peek(int la[])//Function to display and return the top element of the stack
{

    if (top == -1)//If the top value is -1, this means that stack is empty and there is nothing to delete. This is called underflow condtion.
    {
        printf("Underflow\n\n");
        return 0;
    }
    else
    {
        printf("The top element is " "%d" "\n\n", la[top]);
        return la[top];
    }
}

void display(int la[])//Function to display the contents of the stack.
{

    if (top >= 0)//If top is greater than equal to 0, ie stack is not empty.
    {
        printf("Stack : ");
        for (int i = top; i >= 0; i--)//Loop to print the stack in reverse order in which elements were added.
        {
            printf("%d ", la[i]);
        }
        printf("\n\n");
    }
    else//If stack is empty
    {
        printf("The stack is empty.\n\n");
    }
}

int main()
{
    int stack[100];//Initialize array for the stack to be made. Specified initialize size to avoid to stackdump.
    int m;//Variable to store the maximum number of elements the stack can have. Fixes the size of stack.
    printf("Enter the no. of elements in the stack (1 - 100) : ");
    scanf(" %d", &m);
    printf("\n");

    int a = 0;
    while (a < 1)//While loop to perform various operations as the user likes.
    {
        int response;
        printf("Stack Operations : Push = 1  Pop = 2  Peek = 3  Display = 4  End = 5\n\n");
        printf("Enter your choice = ");
        scanf(" %d", &response);
        switch (response)//Switch case, works depending on the input by user.
        {
        case 1:
        {
            push(stack, m);
            break;
        }

        case 2:
        {
            pop(stack);
            break;
        }

        case 3:
        {
            peek(stack);
            break;
        }

        case 4:
        {
            display(stack);
            break;
        }

        case 5:
        {
            a++;
            break;
        }

        default:
        {
            printf("Invalid input. Please try again.\n\n");
            break;
        }
        }
    }
    printf("\n");

    return 0;
}

/* Size of stack (m) = 6

Pushed the elements in this order = 1 2 3 4 5 6(top)

Display = 6 5 4 3 2 1

Pop = Removed 6
Pop = Removed 5

Peek = Displayed and returned 4

Display = 4 3 2 1

Time Complexity of 
1.Push, Pop, Peek = O(1)
2.Display = O(n)
*/

File: Kotlin/Stack.kt
import java.util.*

/**
 * Stack Implementation which supports fixed-size or dynamic
 * Operations && Time Complexities:
 *    Push - O(1) -- adds new element to the stack
 *    Pop - O(1)  -- removes the top element from the stack and returns the removed element,
 *    isEmpty - O(1) -- returns true if stack is empty
 *    peek - O(1) -- returns top element of the stack
 *    print - O(size) -- prints elements present in the stack
 */
class Stack(private val size: Int = 0) {
    private var maxStackSize = if (size > 0) size else 65536 // 2 ^ 16 = 65536 is chosen to save memory.
    private var array = IntArray(maxStackSize)
    private var stackPointer = -1;
    fun push(value: Int) {
        if (++stackPointer == maxStackSize) {
            stackPointer -= 1
            println("Max $size Limit Reached ..OVERFLOW");
            return;
        }
        array[stackPointer] = value
    }

    fun peek(): Int {
        if (stackPointer < 0) {
            println("No Element present in stack")
            return -1;
        }
        return array[stackPointer]
    }

    fun pop(): Int {
        if (stackPointer < 0) {
            println("No Element present in stack..UNDERFLOW")
            return -1;
        }
        println("${array[stackPointer]} is popped.")
        return array[stackPointer--];
    }

    fun print() {
        for (index in stackPointer downTo 0) {
            print("${array[index]} ")
        }
        println()
    }

    fun isEmpty(): Boolean {
        return stackPointer == -1;
    }
}

/**
Example:
Enter size for fixed size stack or click enter for dynamic stack
5
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
1
Enter element to be pushed
1
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
1
Enter element to be pushed
2
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
1
Enter element to be pushed
3
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
1
Enter element to be pushed
4
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
1
Enter element to be pushed
5
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
1
Enter element to be pushed
6
Max 5 Limit Reached ..OVERFLOW
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
3
Element at the top of the stack is 5
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
4
5 4 3 2 1
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
5
false
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
2
5 is popped.
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
2
4 is popped.
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
3
Element at the top of the stack is 3
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
5
false
Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6
Enter your choice =
6
 **/
fun main() {
    val scanner = Scanner(System.`in`);
    println("Enter size for fixed size stack or click enter for dynamic stack");
    val size: Int = readLine()?.toIntOrNull() ?: 0;
    val stack = Stack(size)
    while (true) {
        println("Stack Operations : Push = 1  Pop = 2  Peek = 3  Print = 4  isEmpty = 5 End = 6");
        println("Enter your choice = ");
        when (scanner.nextInt()) {
            1 -> {
                println("Enter element to be pushed")
                stack.push(scanner.nextInt())
            }
            2 -> {
                stack.pop()
            }
            3 -> {
                println(stack.peek())
            }
            4 -> {
                stack.print()
            }
            5 -> {
                println(stack.isEmpty())
            }
            6 -> {
                break;
            }
        }
    }
}
